### PR TITLE
maint: disable implicit conversion of 0 to handle

### PR DESCRIPTION
**Pull request by Skylion007**
## Description
* Disable implicit conversion of int to pyobject* to handle.

## Suggested changelog entry:

```rst

```

### CODE REVIEW

File: include/pybind11/pytypes.h [hunk 1]
```diff
@@ -214,6 +214,7 @@ class handle : public detail::object_api<handle> {
     /// Creates a ``handle`` from the given raw Python object pointer
     // NOLINTNEXTLINE(google-explicit-constructor)
     handle(PyObject *ptr) : m_ptr(ptr) {} // Allow implicit conversion from PyObject*
```
Comment (rwgk) [inline include/pybind11/pytypes.h:216]: Maybe this is a job for `enable_if`?
I'd bet though `enable_if` `PyObject*` won't work, `is_pointer` or similar might. Or maybe to be fancy, look for `ob_refcnt` in the pointee?

Comment (Skylion007) [reply include/pybind11/pytypes.h:216]: @rwgk Hmm... can't figure out the way to construct the enable_if_t.

Comment (Skylion007) [reply include/pybind11/pytypes.h:216]: This strategy doesn't work because it causes ambiguous overloads, which is ridiculous since the func is deleted.

**Comment by rwgk** [comment]
It's a can of worms indeed, but I got pretty far:

```
rwgk.c.googlers.com:~/forked/pybind11 $ git diff
diff --git a/include/pybind11/pytypes.h b/include/pybind11/pytypes.h
index f9625e77..370b3f9a 100644
--- a/include/pybind11/pytypes.h
+++ b/include/pybind11/pytypes.h
@@ -213,7 +213,13 @@ public:
     handle() = default;
     /// Creates a ``handle`` from the given raw Python object pointer
     // NOLINTNEXTLINE(google-explicit-constructor)
-    handle(PyObject *ptr) : m_ptr(ptr) {} // Allow implicit conversion from PyObject*
+    template <typename T, detail::enable_if_t<std::is_pointer<T>::value || std::is_same<T, std::nullptr_t>::value, int> = 0>
+    handle(T ptr) : m_ptr(ptr) {} // Allow implicit conversion from PyObject*
+
+    handle(const handle &) = default;
+    handle(handle &&) = default;
+    handle& operator=(const handle &) = default;
+    handle& operator=(handle &&) = default;

     /// Return the underlying ``PyObject *`` pointer
     PyObject *ptr() const { return m_ptr; }
diff --git a/tests/test_numpy_array.cpp b/tests/test_numpy_array.cpp
index 69ddbe1e..f8a777e4 100644
--- a/tests/test_numpy_array.cpp
+++ b/tests/test_numpy_array.cpp
@@ -218,7 +218,7 @@ TEST_SUBMODULE(numpy_array, sm) {
     // test_empty_shaped_array
     sm.def("make_empty_shaped_array", [] { return py::array(py::dtype("f"), {}, {}); });
     // test numpy scalars (empty shape, ndim==0)
-    sm.def("scalar_int", []() { return py::array(py::dtype("i"), {}, {}, &data_i); });
+    sm.def("scalar_int", []() { return py::array(py::dtype("i"), {}, {}/*, &data_i*/); });

     // test_wrap
     sm.def("wrap", [](const py::array &a) {
```

There is only one subtest that does not pass (test_make_empty_shaped_array) with this diff. Is there a chance you could help debugging that?

**Comment by Skylion007** [comment]
Closing in favor of #4008

**Pull request closed without merging**